In [345]:
import pandas as pd
import numpy as np

In [79]:
df1 = pd.read_csv(r'D:\Prathinav K V\GreatLearning\Recommendation Systems\phone_user_review_file_1.csv')
df2 = pd.read_csv(r'D:\Prathinav K V\GreatLearning\Recommendation Systems\phone_user_review_file_2.csv')
df3 = pd.read_csv(r'D:\Prathinav K V\GreatLearning\Recommendation Systems\phone_user_review_file_3.csv')
df4 = pd.read_csv(r'D:\Prathinav K V\GreatLearning\Recommendation Systems\phone_user_review_file_4.csv')
df5 = pd.read_csv(r'D:\Prathinav K V\GreatLearning\Recommendation Systems\phone_user_review_file_5.csv')
df6 = pd.read_csv(r'D:\Prathinav K V\GreatLearning\Recommendation Systems\phone_user_review_file_6.csv')

In [80]:
df1.shape,df2.shape,df3.shape,df4.shape,df5.shape,df6.shape

((374910, 11),
 (114925, 11),
 (312961, 11),
 (98284, 11),
 (350216, 11),
 (163837, 11))

In [81]:
df1.head()

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   

                                             extract       author  \
0  As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...    james0923   
2  Adequate feel. Nice heft. Processor's still sl...     R. Craig   
3  Never disappointed. One of the reasons I've be...  Buster2020    
4  I've now found that i'm in a group of people t...   S Ate Mine   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8

In [82]:
df2.head()

phone_url       date lang country  source        domain  \
0  /cellphones/leagoo-lead-7/  4/15/2015   en      us  Amazon    amazon.com   
1  /cellphones/leagoo-lead-7/  5/23/2015   en      gb  Amazon  amazon.co.uk   
2  /cellphones/leagoo-lead-7/  4/27/2015   en      gb  Amazon  amazon.co.uk   
3  /cellphones/leagoo-lead-7/  4/22/2015   en      gb  Amazon  amazon.co.uk   
4  /cellphones/leagoo-lead-7/  4/18/2015   en      gb  Amazon  amazon.co.uk   

   score  score_max                                            extract  \
0    2.0       10.0  The telephone headset is of poor quality , not...   
1   10.0       10.0  This is my first smartphone so I have nothing ...   
2    8.0       10.0  Great phone. Battery life not great but seems ...   
3   10.0       10.0      Best 90 quid I've ever spent on a smart phone   
4   10.0       10.0  I m happy with this phone.it s very good.thx team   

          author                                            product  
0           luis  Leagoo Lead7 5.0 Inch HD JDI LTPS Screen 3G Sm...  
1     Mark Lavin  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
2         tracey  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
3  Reuben Ingram  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
4         viorel  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...

In [83]:
df3.head()

phone_url       date lang country  \
0  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  11/7/2015   pt      br   
1  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  10/2/2015   pt      br   
2  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
3  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
4  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/1/2015   pt      br   

      source            domain  score  score_max  \
0  Submarino  submarino.com.br    6.0       10.0   
1  Submarino  submarino.com.br   10.0       10.0   
2  Submarino  submarino.com.br   10.0       10.0   
3  Submarino  submarino.com.br    8.0       10.0   
4    Colombo    colombo.com.br    8.0       10.0   

                                             extract  \
0  recomendo, eu comprei um, a um ano, e agora co...   
1  Comprei um pouco desconfiada do site e do celu...   
2  Muito bom o produto, obvio que tem versões mel...   
3  Unica ressalva fica para a camera que poderia ...   
4  Rapidez e atenção na entrega. O aparelho é mui...   

                           author  \
0                herlington tesch   
1             Luisa Silva Marieta   
2                           Cyrus   
3       Marcela Santa Clara Brito   
4  Claudine Maria Kuhn Walendorff   

                                             product  
0  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
1  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
2  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
3  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
4  Smartphone Samsung Galaxy S3 Slim, Dual Chip, ...

In [84]:
df4.head()

phone_url        date lang country  \
0  /cellphones/samsung-s7262-duos-galaxy-ace/   3/11/2015   en      us   
1  /cellphones/samsung-s7262-duos-galaxy-ace/  17/11/2015   en      in   
2  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
3  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
4  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   

   source      domain  score  score_max  \
0  Amazon  amazon.com    2.0       10.0   
1  Zopper  zopper.com   10.0       10.0   
2  Amazon   amazon.in    4.0       10.0   
3  Amazon   amazon.in    6.0       10.0   
4  Amazon   amazon.in   10.0       10.0   

                                             extract            author  \
0  was not conpatable with my phone as stated. I ...  Frances DeSimone   
1  Decent Functions and Easy to Operate Pros:- Th...     Expert Review   
2  Not Good Phone such price. Hang too much and v...   Amazon Customer   
3                               not bad for features   Amazon Customer   
4                                  Excellent product               NHK   

                                             product  
0  Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce...  
1                Samsung Galaxy Star Pro S7262 Black  
2  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
3  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
4  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)

In [85]:
df5.head()

phone_url       date lang country      source  \
0  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
1  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
2  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
3  /cellphones/karbonn-k1616/  4/25/2014   en      in     Naaptol   
4  /cellphones/karbonn-k1616/  4/23/2013   en      in     Naaptol   

          domain  score  score_max  \
0  91mobiles.com    2.0       10.0   
1  91mobiles.com    6.0       10.0   
2  91mobiles.com    4.0       10.0   
3    naaptol.com   10.0       10.0   
4    naaptol.com   10.0       10.0   

                                             extract           author  \
0  I bought 1 month before. currently speaker is ...        venkatesh   
1  I just bought one week back, I have Airtel con...           Venkat   
2  one problem in this handset opera is not worki...           krrish   
3  here Karbonn comes up with an another excellen...  BRIJESH CHAUHAN   
4  What a phone, all so on Naaptol my god 23% off...    Suraj CHAUHAN   

                 product  
0          Karbonn K1616  
1          Karbonn K1616  
2          Karbonn K1616  
3  Karbonn K1616 - Black  
4  Karbonn K1616 - Black

In [86]:
df6.head()

phone_url        date lang country  \
0  /cellphones/samsung-instinct-sph-m800/   9/16/2011   en      us   
1  /cellphones/samsung-instinct-sph-m800/   2/13/2014   en      us   
2  /cellphones/samsung-instinct-sph-m800/  12/30/2011   en      us   
3  /cellphones/samsung-instinct-sph-m800/  10/18/2008   en      us   
4  /cellphones/samsung-instinct-sph-m800/    9/6/2008   en      us   

          source             domain  score  score_max  \
0    Phone Arena     phonearena.com    8.0       10.0   
1         Amazon         amazon.com    6.0       10.0   
2    Phone Scoop     phonescoop.com    9.0       10.0   
3  HandCellPhone  handcellphone.com    4.0       10.0   
4   Reviewed.com       reviewed.com    6.0       10.0   

                                             extract      author  \
0  I've had the phone for awhile and it's a prett...  ajabrams95   
1  to be clear it is not the sellers fault that t...   Stephanie   
2  Well i love this phone. i have had ton of phon...    snickers   
3  I have had my Instinct for several months now ...         A4C   
4  i have had this instinct phone for about two m...   betaBgood   

                     product  
0        Samsung Instinct HD  
1  Samsung SPH M800 Instinct  
2              Instinct M800  
3           Samsung Instinct  
4           Samsung Instinct

In [87]:
df = pd.concat([df1,df2,df3,df4,df5,df6])

In [88]:
df.shape

(1415133, 11)

In [89]:
df.sample(50)

phone_url        date lang  \
71032                    /cellphones/samsung-sgh-z540/   9/30/2008   ru   
111595                           /cellphones/nokia-n8/    2/9/2011   en   
297247                    /cellphones/nokia-lumia-920/   3/13/2013   en   
296337                           /cellphones/lg-k500n/   1/26/2017   nl   
362546                   /cellphones/motorola-moto-g4/   5/18/2016   es   
83658                     /cellphones/htc-desire-326g/   1/13/2016   ru   
42135              /cellphones/sony-xperia-z3-compact/   1/24/2016   de   
75923             /cellphones/samsung-galaxy-note-iii/   1/10/2014   de   
192171                           /cellphones/nokia-c5/   8/19/2012   de   
307394                   /cellphones/nokia-3600-slide/    6/7/2009   de   
56033                /cellphones/huawei-ascend-mate-8/   11/5/2016   de   
347440                  /cellphones/samsung-galaxy-s5/    6/7/2014   en   
315079                /cellphones/samsung-galaxy-s-ii/  12/10/2013   en   
195194                         /cellphones/nokia-5230/    2/9/2012   de   
276129                           /cellphones/lg-lx370/    8/9/2009   en   
76165                    /cellphones/samsung-sgh-u600/   9/29/2010   fr   
31735                 /cellphones/samsung-galaxy-note/   25/2/2012   ru   
38330                    /cellphones/samsung-sph-m390/   11/1/2012   en   
160163                           /cellphones/nokia-e7/  11/14/2011   de   
49253                      /cellphones/apple-iphone-5/  28/12/2012   en   
100729                /cellphones/samsung-m300-136513/    8/9/2010   en   
133898             /cellphones/samsung-galaxy-prevail/   6/12/2011   en   
40955                     /cellphones/apple-iphone-6s/   4/19/2017   en   
91430              /cellphones/samsung-galaxy-a5-2017/   4/30/2017   de   
145493                  /cellphones/lg-e900-optimus-7/   2/20/2011   de   
147761           /cellphones/samsung-galaxy-s7-active/    4/7/2017   en   
121118          /cellphones/samsung-champ-duos-e2652w/    8/3/2012   tr   
51219           /cellphones/sony-ericsson-spiro-w100i/    4/5/2013   de   
42272                     /cellphones/apple-iphone-6s/  11/24/2015   en   
17444                        /cellphones/samsung-j210/    8/7/2011   ru   
71541                   /cellphones/lg-google-nexus-5/   11/3/2014   ru   
129673  /cellphones/motorola-i730-i733-i710-i720-i740/    4/3/2004   en   
192823     /cellphones/samsung-galaxy-s5-neo-sm-g903f/  11/23/2015   de   
78887                          /cellphones/blu-win-jr/   4/19/2015   en   
86837                         /cellphones/nokia-c2-05/    1/8/2014   it   
50298                 /cellphones/lenovo-vibe-k4-note/    4/6/2016   en   
96204                     /cellphones/nokia-lumia-610/   9/19/2014   it   
289423                           /cellphones/lg-kf690/    3/2/2009   ru   
287092                     /cellphones/blu-studio-5-0/  12/30/2014   en   
105589                             /cellphones/lg-l70/   8/12/2014   pt   
112919                       /cellphones/xiaomi-mi-4i/    7/2/2015   en   
139254                     /cellphones/htc-one-mini-2/  11/10/2014   de   
79811                     /cellphones/apple-iphone-4s/   16/1/2012   en   
280062                          /cellphones/nokia-206/  11/27/2014   it   
250698           /cellphones/samsung-galaxy-s-ii-plus/   2/16/2014   it   
196355                      /cellphones/sony-xperia-l/   8/26/2013   ru   
323498             /cellphones/samsung-sgh-i900-omnia/   10/1/2012   ru   
41751                     /cellphones/apple-iphone-5s/    1/2/2016   en   
227451                         /cellphones/htc-one-m9/   5/16/2015   en   
342560                  /cellphones/samsung-galaxy-s5/   6/29/2015   en   

       country         source              domain  score  score_max  \
71032       ru         Yandex    market.yandex.ru   10.0       10.0   
111595      gb         Amazon        amazon.co.uk    2.0       10.0   
297247

In [90]:
df.lang.unique()

array(['en', 'he', 'fi', 'it', 'no', 'nl', 'ru', 'cs', 'sv', 'es', 'pt',
       'fr', 'de', 'da', 'tr', 'pl', 'hu', 'ar', 'id', 'zh', 'ja', 'ko'],
      dtype=object)

In [91]:
df.lang.nunique()

22

In [92]:
df.country.unique()

array(['us', 'gb', 'au', 'in', 'il', 'fi', 'it', 'no', 'nl', 'ru', 'cz',
       'be', 'se', 'es', 'br', 'fr', 'ua', 'ca', 'de', 'dk', 'nz', 'tr',
       'pl', 'ch', 'pt', 'hu', 'za', 'mx', 'ae', 'id', 've', 'sg', 'cl',
       'ar', 'ec', 'tw', 'co', 'jp', 'uy', 'pe', 'cn', 'kr'], dtype=object)

In [93]:
df.country.nunique()

42

In [94]:
df.source.nunique()

331

In [95]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1415133 entries, 0 to 163836
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   phone_url  1415133 non-null  object 
 1   date       1415133 non-null  object 
 2   lang       1415133 non-null  object 
 3   country    1415133 non-null  object 
 4   source     1415133 non-null  object 
 5   domain     1415133 non-null  object 
 6   score      1351644 non-null  float64
 7   score_max  1351644 non-null  float64
 8   extract    1395772 non-null  object 
 9   author     1351931 non-null  object 
 10  product    1415132 non-null  object 
dtypes: float64(2), object(9)
memory usage: 129.6+ MB


In [96]:
df.isna().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        63489
score_max    63489
extract      19361
author       63202
product          1
dtype: int64

In [97]:
((df.isna().sum())/len(df))*100

phone_url    0.000000
date         0.000000
lang         0.000000
country      0.000000
source       0.000000
domain       0.000000
score        4.486433
score_max    4.486433
extract      1.368140
author       4.466153
product      0.000071
dtype: float64

In [98]:
df[df.duplicated()]

phone_url        date lang country  \
342              /cellphones/samsung-galaxy-s8/   4/26/2017   nl      nl   
343              /cellphones/samsung-galaxy-s8/   4/26/2017   nl      nl   
371              /cellphones/samsung-galaxy-s8/   4/25/2017   nl      nl   
452              /cellphones/samsung-galaxy-s8/   4/25/2017   nl      nl   
3718    /cellphones/samsung-galaxy-s6-edgeplus/   10/4/2015   en      us   
...                                         ...         ...  ...     ...   
146836                  /cellphones/nokia-8850/   8/31/2010   ru      ru   
146838                  /cellphones/nokia-8850/   6/25/2010   ru      ru   
158911             /cellphones/samsung-sgh-600/   4/30/2015   tr      tr   
160164          /cellphones/philips-genie-2000/   8/20/2000   it      it   
161654              /cellphones/motorola-m3888/  12/11/2000   it      it   

          source            domain  score  score_max  \
342     Coolblue       coolblue.nl    9.6       10.0   
343     Coolblue       coolblue.nl    9.6       10.0   
371     Coolblue       coolblue.nl    9.6       10.0   
452     Coolblue       coolblue.nl   10.0       10.0   
3718      Amazon        amazon.com   10.0       10.0   
...          ...               ...    ...        ...   
146836    Yandex  market.yandex.ru   10.0       10.0   
146838    Yandex  market.yandex.ru   10.0       10.0   
158911  Cepworld      cepworld.com    NaN        NaN   
160164      Ciao           ciao.it    8.0       10.0   
161654      Ciao           ciao.it   10.0       10.0   

                                                  extract             author  \
342     Ik vind het tot nu toe een heel erg aangenaam ...  Stefanie Hinneman   
343     Ik vind het tot nu toe een heel erg aangenaam ...  Stefanie Hinneman   
371     Tot nu toe in de eerste dagen weinig tot geen ...     Robin Winckens   
452                                  Wat een huge scherm!            B Yapar   
3718    For readers, my opinions on this device (ATT 3...        FrozenCloud   
...                                                   ...                ...   
146836  Просто отличный телефон, который приятно лежит...                NaN   
146838  За свои деньги, 1000 долларов которые он когда...                NaN   
158911                            varsa satan 05312614607             muhsin   
160164  Il Genie e' un buon telefono, scorrimente del ...          tonyblack   
161654  E' un telefonino qualitativamente ottimo sopra...                NaN   

                                                  product  
342                               Samsung Galaxy S8 Zwart  
343                               Samsung Galaxy S8 Zwart  
371                               Samsung Galaxy S8 Zwart  
452                               Samsung Galaxy S8 Zwart  
3718    Samsung Galaxy S6 Edge+, Black 64GB (Verizon W...  
...                                                   ...  
146836                                         Nokia 8850  
146838                                         Nokia 8850  
158911                              Samsung (936) SGH-600  
160164                                 Philips Genie 2000  
161654                                     Motorola M3888  

[6412 rows x 11 columns]

### 1. The data has reviews in 22 different languages, from 42 different countries, and from 331 different sources.
### 2. The data has missing values and duplicate values. 
### 3. The data contains other information which is not very useful like date , phone url, and score_max

In [99]:
df.score_max.unique()

array([10., nan])

In [100]:
# Rounding off the data to the nearest integer

In [101]:
df = df.round()

In [102]:
#Imputing missing values

In [103]:
df.isna().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        63489
score_max    63489
extract      19361
author       63202
product          1
dtype: int64

In [104]:
round(df['score'].mean())

8

In [105]:
df['score'].fillna(round(df['score'].mean()),inplace=True)

In [106]:
df.isna().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score            0
score_max    63489
extract      19361
author       63202
product          1
dtype: int64

In [107]:
df['extract'].fillna('-',inplace=True)

In [108]:
df[df['author'].isna()]

phone_url       date lang country  \
270              /cellphones/samsung-galaxy-s8/   5/4/2017   cs      cz   
456              /cellphones/samsung-galaxy-s8/  4/25/2017   cs      cz   
548     /cellphones/samsung-galaxy-s6-edgeplus/  2/13/2016   en      us   
593     /cellphones/samsung-galaxy-s6-edgeplus/  12/1/2016   en      us   
4549    /cellphones/samsung-galaxy-s6-edgeplus/  6/16/2016   ru      ru   
...                                         ...        ...  ...     ...   
163671             /cellphones/ericsson-gf-768/   1/9/2000   de      de   
163706             /cellphones/ericsson-pf-768/  1/28/2000   de      de   
163708             /cellphones/ericsson-pf-768/  1/13/2000   de      de   
163713              /cellphones/motorola-m3288/  7/30/2001   it      it   
163800                 /cellphones/siemens-c10/   1/7/2000   de      de   

         source            domain  score  score_max  \
270         CZC            czc.cz   10.0       10.0   
456         CZC            czc.cz   10.0       10.0   
548         HSN           hsn.com    8.0        NaN   
593     Samsung       samsung.com   10.0       10.0   
4549     Yandex  market.yandex.ru    2.0       10.0   
...         ...               ...    ...        ...   
163671   Dooyoo         dooyoo.de    6.0       10.0   
163706   Dooyoo         dooyoo.de    6.0       10.0   
163708   Dooyoo         dooyoo.de    8.0       10.0   
163713     Ciao           ciao.it    6.0       10.0   
163800   Dooyoo         dooyoo.de    6.0       10.0   

                                                  extract author  \
270     Rozhodně nejlepší smartphone pro rok 2017, vyt...    NaN   
456     Samsung Galaxy S8 G950F 64GB – ÚVOD V poslední...    NaN   
548     The best smart phone I ever purchased. Easy to...    NaN   
593     I Love EVERYTHING ABOUT MY PHONE!! The camera ...    NaN   
4549    Пользуюсь galaxy s6 edge plus кроме понтов тел...    NaN   
...                                                   ...    ...   
163671  Das Ericsson 728 ist ensich schon ein gutes Ge...    NaN   
163706  Aufmerksamkeit erregt das PF 768 mit seinen kr...    NaN   
163708  Aufmerksamkeit erregt das PF 768 mit seinen kr...    NaN   
163713  Evviva la tecnologia!!!! Essa sta avanzando se...    NaN   
163800  Das praktische Easy-Modell für \n ganz normale...    NaN   

                                                  product  
270                        Samsung Galaxy S8, 64GB, černá  
456                        Samsung Galaxy S8, 64GB, černá  
548     Samsung Galaxy S6 Edge+ 32GB Unlocked GSM Andr...  
593                   Samsung Galaxy S6 edge+ 64GB (AT&T)  
4549                         Samsung Galaxy S6 Edge+ 32Gb  
...                                                   ...  
163671                                     Ericsson GF768  
163706                                     Ericsson PF768  
163708                                     Ericsson PF768  
163713                                     Motorola M3288  
163800                                      Siemens C10 D  

[63202 rows x 11 columns]

In [109]:
df['author'].fillna('-',inplace=True)

In [110]:
df.iloc[[270,456]]

phone_url       date lang country source  domain  \
270  /cellphones/samsung-galaxy-s8/   5/4/2017   cs      cz    CZC  czc.cz   
456  /cellphones/samsung-galaxy-s8/  4/25/2017   cs      cz    CZC  czc.cz   

     score  score_max                                            extract  \
270   10.0       10.0  Rozhodně nejlepší smartphone pro rok 2017, vyt...   
456   10.0       10.0  Samsung Galaxy S8 G950F 64GB – ÚVOD V poslední...   

    author                         product  
270      -  Samsung Galaxy S8, 64GB, černá  
456      -  Samsung Galaxy S8, 64GB, černá

In [111]:
df.isna().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score            0
score_max    63489
extract          0
author           0
product          1
dtype: int64

In [112]:
df[df['product'].isna()]

phone_url       date lang country  source  \
312960  /cellphones/samsung-galaxy-s-iii/  1/22/2014   de      de  Amazon   

           domain  score  score_max  \
312960  amazon.de   10.0       10.0   

                                                  extract author product  
312960  Bestes Smartphone was ich bisher hatte :) öafk...      -     NaN

In [119]:
df['product'].dropna(inplace=True)

In [120]:
df.isna().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score            0
score_max    63489
extract          0
author           0
product          1
dtype: int64

In [125]:
df.drop('score_max', axis = 1 , inplace=True)   # Removing score_max to perform dropna on the whole DataFrame

In [126]:
df.isna().sum()

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
extract      0
author       0
product      1
dtype: int64

In [127]:
df.dropna(inplace=True)

In [128]:
df.isna().sum()

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
extract      0
author       0
product      0
dtype: int64

### There are no missing values in the dataframe

In [129]:
# Removing duplicate rows

In [130]:
df.duplicated().sum()

6438

In [134]:
df.drop_duplicates(inplace = True)

In [135]:
df.duplicated().sum()

0

In [136]:
# Sampling 1 million data samples

In [138]:
df = df.sample(n=1000000,random_state=612)

In [139]:
df.shape

(1000000, 10)

In [142]:
dfc1 = df.copy()

In [140]:
# Removing unecessary columns

In [141]:
df.columns

Index(['phone_url', 'date', 'lang', 'country', 'source', 'domain', 'score',
       'extract', 'author', 'product'],
      dtype='object')

In [145]:
df.drop(['phone_url', 'date','source', 'domain'],axis = 1 ,inplace=True)

In [146]:
df.columns

Index(['lang', 'country', 'score', 'extract', 'author', 'product'], dtype='object')

### ____________Identifying the most rated feature_____________

### From the data given it is clear that, the only understandable language is English in the extracts column. 
### Hence we will only use the english extracts to find the most rated features

In [148]:
dfc2 = df.copy()  # making a copy of the data

In [216]:
df_rated = df[df['lang']=='en'].reset_index()

In [217]:
df_rated.shape

(393563, 7)

In [218]:
df_rated

index lang country  score  \
0        55265   en      in    8.0   
1        97318   en      us    2.0   
2       104951   en      gb    8.0   
3       146543   en      in    2.0   
4       194696   en      us   10.0   
...        ...  ...     ...    ...   
393558  135959   en      us    9.0   
393559    9813   en      in    2.0   
393560   43605   en      us   10.0   
393561   93107   en      us   10.0   
393562   43202   en      gb    2.0   

                                                  extract  \
0       K900 had advance features compared to K4 note....   
1       Memory is awful, all this junk preloaded you c...   
2       I purchased the Sony Xperia Z Ultra as a repla...   
3       First android update failed, it had to be comp...   
4       smarphone simply unique, very satisfied with t...   
...                                                   ...   
393558  I love every features for $200. The purpose wh...   
393559  Its not worth buying this product i bought it ...   
393560  There are two outstanding new features on the ...   
393561  SHORT VERSION: The Motorola nexus 6 is a great...   
393562  A very good phone ABSOLUTELY RUINED by the ina...   

                        author  \
0       Vinod Kumar Chengespur   
1                       Sharon   
2                  An Sionnach   
3                  Jomine Jose   
4                Marcos Rivero   
...                        ...   
393558            Unmesh Patel   
393559         Amazon Customer   
393560                   LeeAB   
393561            Tarek khoury   
393562           Mr. N. Bonsor   

                                                  product  
0                        Lenovo Vibe K4 Note (White,16GB)  
1       HTC Desire 816 Black (Virgin mobile) - 5.5 inc...  
2       Xperia Sony Xperia Z Ultra SIM-free Android Sm...  
3                                    Lava Iris X5 (Black)  
4       Sony Xperia L C2105 White (Factory Unlocked) 4...  
...                                                   ...  
393558                                           LG F9100  
393559  Lenovo PHAB Plus Tablet (6.8 inch, 32GB, Wi-Fi...  
393560                                    Apple iPhone 6s  
393561  Motorola Nexus 6 Unlocked Cellphone, 64GB, Mid...  
393562              Motorola Gleam+ Sim-Free Mobile Phone  

[393563 rows x 7 columns]

In [158]:
from sklearn.feature_extraction.text import CountVectorizer
vect = CountVectorizer(analyzer='word',ngram_range=(1,2),stop_words='english', min_df = 0.002)

In [159]:
vect.fit(df_rated['extract'])
title_matrix = vect.transform(df_rated['extract'])

In [160]:
title_matrix.shape

(393563, 1302)

In [161]:
features = vect.get_feature_names()
features

D:\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['00',
 '10',
 '100',
 '12',
 '13',
 '15',
 '16',
 '16gb',
 '1st',
 '20',
 '200',
 '2015',
 '2016',
 '24',
 '2gb',
 '2nd',
 '30',
 '32gb',
 '3g',
 '3rd',
 '4g',
 '4g lte',
 '4s',
 '50',
 '5s',
 '6s',
 '8gb',
 '99',
 'ability',
 'able',
 'absolutely',
 'absolutely love',
 'access',
 'accessories',
 'account',
 'activate',
 'activated',
 'active',
 'actually',
 'add',
 'added',
 'advertised',
 'affordable',
 'ago',
 'allow',
 'alot',
 'amazing',
 'amazing phone',
 'android',
 'android phone',
 'android phones',
 'annoying',
 'answer',
 'anymore',
 'apart',
 'app',
 'apple',
 'application',
 'applications',
 'apps',
 'area',
 'arrived',
 'ask',
 'asus',
 'att',
 'attractive',
 'audio',
 'auto',
 'automatically',
 'available',
 'average',
 'away',
 'awesome',
 'awesome phone',
 'awful',
 'backup',
 'bad',
 'bad phone',
 'bar',
 'barely',
 'based',
 'basic',
 'basic phone',
 'basically',
 'battery',
 'battery backup',
 'battery good',
 'battery lasts',
 'battery life',
 'bb',
 'beast',
 'be

In [223]:
word_count1 = [df_rated["extract"].str.contains(' '+x+' ', case=False).sum() for x in features]
word_count1

[3,
 3207,
 540,
 900,
 762,
 1059,
 925,
 809,
 891,
 923,
 278,
 363,
 443,
 626,
 716,
 934,
 1043,
 708,
 3020,
 828,
 3467,
 543,
 920,
 334,
 979,
 877,
 654,
 40,
 947,
 5105,
 4130,
 1343,
 1738,
 666,
 556,
 981,
 735,
 621,
 3671,
 1574,
 958,
 977,
 971,
 3815,
 781,
 1244,
 7459,
 599,
 13810,
 1613,
 587,
 876,
 778,
 314,
 944,
 3380,
 4251,
 704,
 1366,
 10171,
 789,
 2449,
 1226,
 892,
 1010,
 612,
 1423,
 695,
 916,
 3279,
 1795,
 1874,
 6950,
 537,
 397,
 3342,
 7614,
 315,
 638,
 871,
 1036,
 4682,
 972,
 824,
 41438,
 2083,
 52,
 1340,
 11286,
 772,
 435,
 998,
 3067,
 1833,
 23239,
 553,
 6326,
 996,
 560,
 15952,
 2,
 430,
 8551,
 698,
 1853,
 968,
 702,
 7135,
 2444,
 4233,
 496,
 1147,
 1135,
 2268,
 996,
 23161,
 24,
 2950,
 4007,
 1549,
 819,
 1832,
 1458,
 2033,
 1414,
 564,
 1493,
 702,
 949,
 2447,
 435,
 541,
 3263,
 1678,
 2217,
 584,
 1472,
 3086,
 2651,
 16979,
 62,
 4409,
 11,
 970,
 1228,
 1401,
 5236,
 746,
 9317,
 31557,
 159,
 61,
 263,
 2448,
 684

In [224]:
np.flip(np.sort(word_count1))[0:200]

array([180831,  59482,  41949,  41438,  36887,  32550,  31557,  30844,
        30146,  30051,  25102,  23239,  23161,  22686,  19883,  19631,
        19311,  19110,  19031,  18775,  17981,  16979,  16933,  16529,
        16219,  15952,  15902,  15484,  15201,  14899,  14516,  14156,
        13819,  13810,  13767,  12384,  12375,  11653,  11286,  11260,
        11218,  11189,  11120,  10960,  10227,  10171,  10080,   9642,
         9380,   9322,   9317,   9271,   9221,   9213,   9187,   9179,
         9153,   9047,   8897,   8745,   8551,   8443,   8240,   8210,
         8180,   8046,   7979,   7921,   7730,   7715,   7614,   7459,
         7403,   7356,   7355,   7351,   7172,   7135,   7068,   7014,
         6979,   6976,   6950,   6847,   6664,   6660,   6643,   6642,
         6500,   6497,   6373,   6326,   6314,   6208,   6205,   6134,
         6079,   6064,   6050,   5870,   5860,   5743,   5723,   5713,
         5599,   5476,   5471,   5381,   5236,   5234,   5206,   5187,
      

In [225]:
top_feat_index = np.flip(np.argsort(word_count1))

In [226]:
feats = []
for i in list(top_feat_index):
    feats.append(features[i])
feats

['phone',
 'good',
 'great',
 'battery',
 'like',
 'just',
 'camera',
 'love',
 'screen',
 'use',
 'really',
 'best',
 'bought',
 'got',
 'new',
 'quality',
 'works',
 'samsung',
 'phones',
 'easy',
 'using',
 'buy',
 'mobile',
 'nice',
 'features',
 'better',
 'does',
 'price',
 'time',
 'used',
 'iphone',
 'product',
 'work',
 'android',
 'life',
 'galaxy',
 'little',
 'months',
 'battery life',
 'need',
 'fast',
 'problem',
 'want',
 'happy',
 'sim',
 'apps',
 'working',
 'say',
 'touch',
 'old',
 'came',
 'did',
 'far',
 'looks',
 'recommend',
 'excellent',
 'thing',
 'nokia',
 'lot',
 'looking',
 'big',
 'day',
 'days',
 'card',
 'smart',
 'year',
 'think',
 'having',
 'make',
 'getting',
 'bad',
 'amazing',
 'sound',
 'purchased',
 'problems',
 'years',
 'display',
 'bit',
 'pretty',
 'size',
 'long',
 'know',
 'awesome',
 'device',
 'cell',
 'memory',
 'wanted',
 'worth',
 'month',
 'look',
 'perfect',
 'best phone',
 'performance',
 'way',
 'smartphone',
 'people',
 'small',
 '

In [227]:
feats[0:200]

['phone',
 'good',
 'great',
 'battery',
 'like',
 'just',
 'camera',
 'love',
 'screen',
 'use',
 'really',
 'best',
 'bought',
 'got',
 'new',
 'quality',
 'works',
 'samsung',
 'phones',
 'easy',
 'using',
 'buy',
 'mobile',
 'nice',
 'features',
 'better',
 'does',
 'price',
 'time',
 'used',
 'iphone',
 'product',
 'work',
 'android',
 'life',
 'galaxy',
 'little',
 'months',
 'battery life',
 'need',
 'fast',
 'problem',
 'want',
 'happy',
 'sim',
 'apps',
 'working',
 'say',
 'touch',
 'old',
 'came',
 'did',
 'far',
 'looks',
 'recommend',
 'excellent',
 'thing',
 'nokia',
 'lot',
 'looking',
 'big',
 'day',
 'days',
 'card',
 'smart',
 'year',
 'think',
 'having',
 'make',
 'getting',
 'bad',
 'amazing',
 'sound',
 'purchased',
 'problems',
 'years',
 'display',
 'bit',
 'pretty',
 'size',
 'long',
 'know',
 'awesome',
 'device',
 'cell',
 'memory',
 'wanted',
 'worth',
 'month',
 'look',
 'perfect',
 'best phone',
 'performance',
 'way',
 'smartphone',
 'people',
 'small',
 '

### From observing the top 200 features 
#### Top features are:
    1. Performance
    2. Battery
    2. Camera
    3. Screen
    4. Quality of the product(build of the product and design)
    5. Price
    6. OS - Android or Apple(iPhone)
    9. Brand

In [229]:
df.columns

Index(['lang', 'country', 'score', 'extract', 'author', 'product'], dtype='object')

In [244]:
top_users = df['author'].value_counts().index

In [247]:
top_users = pd.DataFrame(top_users,columns=['Author'])

In [263]:
df['author'].value_counts().values

array([54630, 43823, 13667, ...,     1,     1,     1], dtype=int64)

In [264]:
top_users['count'] = df['author'].value_counts().values

In [265]:
top_users

Author  count
0       Amazon Customer  54630
1                     -  43823
2        Cliente Amazon  13667
3                 e-bit   5967
4       Client d'Amazon   5493
...                 ...    ...
596802       guest-star      1
596803       F. Gerling      1
596804   kirstenmaskell      1
596805         lexx.182      1
596806   marco ghilardi      1

[596807 rows x 2 columns]

In [272]:
top_users[top_users['Author']!='-'].iloc[0:50]

Author  count
0   Amazon Customer  54630
2    Cliente Amazon  13667
3             e-bit   5967
4   Client d'Amazon   5493
5      Amazon Kunde   3291
6         Anonymous   1924
7      einer Kundin   1890
8      einem Kunden   1350
9           unknown   1199
10       Anonymous    1014
11        Александр    738
12            David    711
13           Сергей    644
14            Marco    626
15             Alex    602
16          Дмитрий    584
17           Andrea    582
18           Андрей    535
19          Алексей    525
20           Daniel    508
21            Chris    489
22         Zákazník    472
23          Michael    462
24          Antonio    457
25        Francesco    426
26       Alessandro    413
27  Kindle Customer    398
28     Amazon-Kunde    390
29             John    379
30          Roberto    379
31            Peter    359
32          Anônimo    350
33          Misafir    348
34       zehrabalta    347
35             Luca    334
36          Alberto    333
37          Anonyme    330
38             Mike    329
39           Carlos    319
40        dilemma21    317
41            Гость    314
42          Евгений    311
43          Stefano    311
44            Fabio    298
45             Jose    296
46           anonym    296
47         Giuseppe    296
48        kuzeykizi    295
49         Владимир    291
50            Денис    287

### From the above we notice that the Authors who have given the highest number of ratings are seen
### Note: the Author : '-' is an imputed value for the data records which had missing values for the author feature 

#### The top 50 authors are:

In [273]:
top_users[top_users['Author']!='-'].iloc[0:50]

Author  count
0   Amazon Customer  54630
2    Cliente Amazon  13667
3             e-bit   5967
4   Client d'Amazon   5493
5      Amazon Kunde   3291
6         Anonymous   1924
7      einer Kundin   1890
8      einem Kunden   1350
9           unknown   1199
10       Anonymous    1014
11        Александр    738
12            David    711
13           Сергей    644
14            Marco    626
15             Alex    602
16          Дмитрий    584
17           Andrea    582
18           Андрей    535
19          Алексей    525
20           Daniel    508
21            Chris    489
22         Zákazník    472
23          Michael    462
24          Antonio    457
25        Francesco    426
26       Alessandro    413
27  Kindle Customer    398
28     Amazon-Kunde    390
29             John    379
30          Roberto    379
31            Peter    359
32          Anônimo    350
33          Misafir    348
34       zehrabalta    347
35             Luca    334
36          Alberto    333
37          Anonyme    330
38             Mike    329
39           Carlos    319
40        dilemma21    317
41            Гость    314
42          Евгений    311
43          Stefano    311
44            Fabio    298
45             Jose    296
46           anonym    296
47         Giuseppe    296
48        kuzeykizi    295
49         Владимир    291
50            Денис    287

In [ ]:
# Finding the products with more than 50 reviews

In [283]:
df['product'].value_counts()

Lenovo Vibe K4 Note (White,16GB)                                                                                                                3709
Lenovo Vibe K4 Note (Black, 16GB)                                                                                                               3083
OnePlus 3 (Graphite, 64 GB)                                                                                                                     2890
OnePlus 3 (Soft Gold, 64 GB)                                                                                                                    2522
Huawei P8lite zwart / 16 GB                                                                                                                     1895
                                                                                                                                                ... 
SAMSUNG E950                                                                                              

In [274]:
top_product = df['product'].value_counts().index

In [278]:
top_product = pd.DataFrame(top_product,columns=['product'])

In [279]:
top_product

product
0                       Lenovo Vibe K4 Note (White,16GB)
1                      Lenovo Vibe K4 Note (Black, 16GB)
2                            OnePlus 3 (Graphite, 64 GB)
3                           OnePlus 3 (Soft Gold, 64 GB)
4                            Huawei P8lite zwart / 16 GB
...                                                  ...
56438                                       SAMSUNG E950
56439                                       Kyocera K404
56440                    Nokia ASHA 206 SIM FREE MAGENTA
56441  LG G2 Mini -(D620R) Smartphone Orange Libero A...
56442                    Sonim XP3300 Force 2" 185g Nero

[56443 rows x 1 columns]

In [280]:
df['product'].value_counts().values

array([3709, 3083, 2890, ...,    1,    1,    1], dtype=int64)

In [281]:
top_product['count'] = df['product'].value_counts().values

In [282]:
top_product

product  count
0                       Lenovo Vibe K4 Note (White,16GB)   3709
1                      Lenovo Vibe K4 Note (Black, 16GB)   3083
2                            OnePlus 3 (Graphite, 64 GB)   2890
3                           OnePlus 3 (Soft Gold, 64 GB)   2522
4                            Huawei P8lite zwart / 16 GB   1895
...                                                  ...    ...
56438                                       SAMSUNG E950      1
56439                                       Kyocera K404      1
56440                    Nokia ASHA 206 SIM FREE MAGENTA      1
56441  LG G2 Mini -(D620R) Smartphone Orange Libero A...      1
56442                    Sonim XP3300 Force 2" 185g Nero      1

[56443 rows x 2 columns]

In [ ]:
# products with over 50 ratings 

In [285]:
top_product[top_product['count'] > 50 ]

product  count
0               Lenovo Vibe K4 Note (White,16GB)   3709
1              Lenovo Vibe K4 Note (Black, 16GB)   3083
2                    OnePlus 3 (Graphite, 64 GB)   2890
3                   OnePlus 3 (Soft Gold, 64 GB)   2522
4                    Huawei P8lite zwart / 16 GB   1895
...                                          ...    ...
4252                                 T-Mobile G1     51
4253  Sony Ericsson K610i Evening Red UMTS Handy     51
4254                  SAMSUNG Player Light Blanc     51
4255                  Sony Xperia C3 Dual (Mint)     51
4256                  Smartphone LG X Power K220     51

[4257 rows x 2 columns]

In [290]:
tph = top_product[top_product['count'] > 50 ]

In [287]:
top_users[(top_users['Author']!='-')&(top_users['count']>50)]

Author  count
0    Amazon Customer  54630
2     Cliente Amazon  13667
3              e-bit   5967
4    Client d'Amazon   5493
5       Amazon Kunde   3291
..               ...    ...
644           Johnny     51
645              Ram     51
646           Jeremy     51
647            Aaron     51
648             José     51

[648 rows x 2 columns]

In [291]:
tau = top_users[(top_users['Author']!='-')&(top_users['count']>50)]

In [289]:
df.shape

(1000000, 6)

### All data which has authors who have given more than 50 ratings and products which have more than 50 ratings

In [294]:
super_df = df[(df['author'].isin(tau['Author']))&(df['product'].isin(tph['product']))]

In [296]:
super_df

lang country  score                                            extract  \
230845   it      it    8.0  Il prodotto e molto valido hardware ottimo l'u...   
15508    en      us    2.0                   Can't download the aps. I wanted   
119096   en      in   10.0  OnePlus 3T is awesome.. It's camer quality, pr...   
345559   en      us    8.0  I just got this phone today and I like it. The...   
102400   en      us    6.0  Awesome mobile, but batery life is really bad ...   
...     ...     ...    ...                                                ...   
161619   ru      ru    8.0  Да как так-то! Почти самая необходимая функция...   
173808   sv      se    8.0  Trevlig smartphone med bra skärm, bra ljud och...   
9813     en      in    2.0  Its not worth buying this product i bought it ...   
101230   ru      ru    6.0  На смартфон не жалуюсь. Отличный, просто супер...   
243653   pt      br   10.0  Um ótimo aparelho, atendendo todas as minhas n...   

                 author                                            product  
230845   Cliente Amazon  Microsoft Telefonia Lumia 950 XL Smartphone, 3...  
15508   Amazon Customer  Nokia Lumia 635 8GB Unlocked GSM 4G LTE Window...  
119096  Amazon Customer       OnePlus 3T (Gunmetal, 6GB RAM + 64GB memory)  
345559        Anonymous                                            LG enV3  
102400  Amazon Customer  Samsung Galaxy S6 G920F 32GB Factory Unlocked ...  
...                 ...                                                ...  
161619          Дмитрий                           Samsung Galaxy A3 (2016)  
173808             Andy                              Samsung Galaxy Xcover  
9813    Amazon Customer  Lenovo PHAB Plus Tablet (6.8 inch, 32GB, Wi-Fi...  
101230           Андрей                           Samsung S5830 Galaxy Ace  
243653            e-bit  Smartphone Motorola Moto X 2ª Geração XT1097 3...  

[101649 rows x 6 columns]

In [295]:
super_df.shape

(101649, 6)

### The shape of the corresponding data is (101649,6)

In [332]:
# Let us consider a dataset which only has products 
# which have more than 1000 reviews and authors who have given more than 1000 reviews

In [330]:
tph = top_product[top_product['count'] > 1000 ]
tau = top_users[(top_users['Author']!='-')&(top_users['count']>1000)]
super_df = df[(df['author'].isin(tau['Author']))&(df['product'].isin(tph['product']))]

In [331]:
super_df

lang country  score                                            extract  \
119096   en      in   10.0  OnePlus 3T is awesome.. It's camer quality, pr...   
21588    en      in    8.0  Very nice performance with this one. Very happ...   
70378    en      in   10.0                                   Marvelous Phone.   
66876    it      it   10.0  Ottimo smartphone, veloce e comodissimo da usa...   
46253    en      in   10.0                         suoer i enjoying very much   
...     ...     ...    ...                                                ...   
366837   en      in   10.0  Pros: Stunning looks, ultimate browsing experi...   
73623    en      in    4.0  Not much so good, poor battery backup not even...   
20129    en      in    8.0  Great battery backup No heating problem Good c...   
69508    en      in   10.0  This phone is very good & good Betty packup an...   
117356   en      in    8.0  This is a good cell phone for browsing and camera   

                 author                                            product  
119096  Amazon Customer       OnePlus 3T (Gunmetal, 6GB RAM + 64GB memory)  
21588   Amazon Customer                Lenovo Vibe K5 (Gold, VoLTE update)  
70378   Amazon Customer                       OnePlus 3 (Soft Gold, 64 GB)  
66876    Cliente Amazon  Huawei P9 Lite Smartphone, LTE, Display 5.2'' ...  
46253   Amazon Customer                     Apple iPhone 5s (Silver, 16GB)  
...                 ...                                                ...  
366837  Amazon Customer                             OnePlus X (Onyx, 16GB)  
73623   Amazon Customer                       OnePlus 3 (Soft Gold, 64 GB)  
20129   Amazon Customer                Lenovo Vibe K5 (Grey, VoLTE update)  
69508   Amazon Customer                 YU Yuphoria YU5010A (Black+Silver)  
117356  Amazon Customer       OnePlus 3T (Gunmetal, 6GB RAM + 64GB memory)  

[14236 rows x 6 columns]

### Popularity Based Filtering 

In [82]:
super_df.groupby('product').agg({'score':'mean'}).reset_index()

product     score
0                      Apple iPhone 5s (Silver, 16GB)  7.030426
1   Huawei P8 Lite Smartphone, Display 5" IPS, Pro...  8.164794
2   Huawei P9 Lite Smartphone, LTE, Display 5.2'' ...  9.043210
3                                            LG GX200  9.000000
4   Lenovo Motorola Moto G Smartphone, 4,5 pollici...  8.636364
5        Lenovo Used Lenovo Zuk Z1 (Space Grey, 64GB)  5.065191
6                   Lenovo Vibe K4 Note (Black, 16GB)  7.049573
7                    Lenovo Vibe K4 Note (White,16GB)  6.973745
8                 Lenovo Vibe K5 (Gold, VoLTE update)  6.553425
9                 Lenovo Vibe K5 (Grey, VoLTE update)  6.403292
10                      Lenovo Vibe K5 (Silver, 16GB)  6.547692
11                Motorola Moto G Turbo (White, 16GB)  6.672932
12                             Nokia 5130 XpressMusic  6.666667
13                             Nokia 5530 XpressMusic  8.000000
14                             Nokia 5800 XpressMusic  8.500000
15                                          Nokia N95  2.000000
16                        OnePlus 3 (Graphite, 64 GB)  8.530739
17                       OnePlus 3 (Soft Gold, 64 GB)  8.313072
18       OnePlus 3T (Gunmetal, 6GB RAM + 64GB memory)  8.309904
19                             OnePlus X (Onyx, 16GB)  7.571125
20                   Samsung Galaxy S5 16GB (Verizon)  9.205882
21                  Samsung Galaxy S7 32GB (T-Mobile)  9.647059
22                   Samsung Galaxy S7 32GB (Verizon)  9.414634
23             Samsung Galaxy S7 edge 32GB (T-Mobile)  9.527273
24              Samsung Galaxy S7 edge 32GB (Verizon)  9.437500
25  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  9.580029
26                 YU Yuphoria YU5010A (Black+Silver)  6.531447

In [83]:
df_grouped = super_df.groupby('product').agg({'score':'mean'}).reset_index()

In [84]:
df_sort = df_grouped.sort_values(['score', 'product'], ascending = False) 
      
#Generate a recommendation rank based upon score 
df_sort['Rank'] = df_sort['score'].rank(ascending=0, method='first') 
          
#Get the top 5 recommendations 
popularity_recommendations = df_sort 
popularity_recommendations.head(10) 

product     score  Rank
21                  Samsung Galaxy S7 32GB (T-Mobile)  9.647059   1.0
25  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  9.580029   2.0
23             Samsung Galaxy S7 edge 32GB (T-Mobile)  9.527273   3.0
24              Samsung Galaxy S7 edge 32GB (Verizon)  9.437500   4.0
22                   Samsung Galaxy S7 32GB (Verizon)  9.414634   5.0
20                   Samsung Galaxy S5 16GB (Verizon)  9.205882   6.0
2   Huawei P9 Lite Smartphone, LTE, Display 5.2'' ...  9.043210   7.0
3                                            LG GX200  9.000000   8.0
4   Lenovo Motorola Moto G Smartphone, 4,5 pollici...  8.636364   9.0
16                        OnePlus 3 (Graphite, 64 GB)  8.530739  10.0

### ^^^ Top 10 products based on Popularity based filtering

In [1]:
import numpy as np
import pandas as pd

In [2]:
# Reading the million dataset
df = pd.read_csv(r'D:\Prathinav K V\GreatLearning\Recommendation Systems\mill_data.csv')
# Reading the super mill dataset
super_df = pd.read_csv(r'D:\Prathinav K V\GreatLearning\Recommendation Systems\super_mill_data.csv')

In [3]:
df.drop(['Unnamed: 0'],axis=1,inplace=True)

In [4]:
df

lang country  score                                            extract  \
0        en      in    8.0  K900 had advance features compared to K4 note....   
1        en      us    2.0  Memory is awful, all this junk preloaded you c...   
2        en      gb    8.0  I purchased the Sony Xperia Z Ultra as a repla...   
3        en      in    2.0  First android update failed, it had to be comp...   
4        en      us   10.0  smarphone simply unique, very satisfied with t...   
...     ...     ...    ...                                                ...   
999995   pt      br   10.0  Um ótimo aparelho, atendendo todas as minhas n...   
999996   fr      fr    8.0  Je suis Tres satifaite par les produits que j'...   
999997   fr      fr    4.0  l'appareil photo est décevant. Couleurs trop s...   
999998   it      it   10.0  Sono rimasto Molto soddisfatto, e come me lo a...   
999999   ru      ru   10.0  Купил себе данный телефон, потому что наконец ...   

                        author  \
0       Vinod Kumar Chengespur   
1                       Sharon   
2                  An Sionnach   
3                  Jomine Jose   
4                Marcos Rivero   
...                        ...   
999995                   e-bit   
999996         Client d'Amazon   
999997                    JYLA   
999998       francesco imbimbo   
999999                  VIcont   

                                                  product  
0                        Lenovo Vibe K4 Note (White,16GB)  
1       HTC Desire 816 Black (Virgin mobile) - 5.5 inc...  
2       Xperia Sony Xperia Z Ultra SIM-free Android Sm...  
3                                    Lava Iris X5 (Black)  
4       Sony Xperia L C2105 White (Factory Unlocked) 4...  
...                                                   ...  
999995  Smartphone Motorola Moto X 2ª Geração XT1097 3...  
999996  Buyus X2 Film Protecteur d'écran VITRE en VERR...  
999997      Samsung Smartphone Samsung Galaxy J7 16 Go Or  
999998  Cubot X9 5.0" IPS HD 3G Smartphone HotKnot And...  
999999                          Sony Ericsson Cedar J108i  

[1000000 rows x 6 columns]

In [5]:
super_df.drop(['Unnamed: 0'],axis=1,inplace=True)
super_df

lang country  score                                            extract  \
0       en      in   10.0  OnePlus 3T is awesome.. It's camer quality, pr...   
1       en      in    8.0  Very nice performance with this one. Very happ...   
2       en      in   10.0                                   Marvelous Phone.   
3       it      it   10.0  Ottimo smartphone, veloce e comodissimo da usa...   
4       en      in   10.0                         suoer i enjoying very much   
...    ...     ...    ...                                                ...   
14231   en      in   10.0  Pros: Stunning looks, ultimate browsing experi...   
14232   en      in    4.0  Not much so good, poor battery backup not even...   
14233   en      in    8.0  Great battery backup No heating problem Good c...   
14234   en      in   10.0  This phone is very good & good Betty packup an...   
14235   en      in    8.0  This is a good cell phone for browsing and camera   

                author                                            product  
0      Amazon Customer       OnePlus 3T (Gunmetal, 6GB RAM + 64GB memory)  
1      Amazon Customer                Lenovo Vibe K5 (Gold, VoLTE update)  
2      Amazon Customer                       OnePlus 3 (Soft Gold, 64 GB)  
3       Cliente Amazon  Huawei P9 Lite Smartphone, LTE, Display 5.2'' ...  
4      Amazon Customer                     Apple iPhone 5s (Silver, 16GB)  
...                ...                                                ...  
14231  Amazon Customer                             OnePlus X (Onyx, 16GB)  
14232  Amazon Customer                       OnePlus 3 (Soft Gold, 64 GB)  
14233  Amazon Customer                Lenovo Vibe K5 (Grey, VoLTE update)  
14234  Amazon Customer                 YU Yuphoria YU5010A (Black+Silver)  
14235  Amazon Customer       OnePlus 3T (Gunmetal, 6GB RAM + 64GB memory)  

[14236 rows x 6 columns]

In [6]:
pip install surprise

  Using cached surprise-0.1-py2.py3-none-any.whl (1.8 kB)
  Using cached scikit-surprise-1.1.1.tar.gz (11.8 MB)
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.1-cp38-cp38-win_amd64.whl size=714219 sha256=3f7c98e1e189adec5a50dc6f139e4f56f54f0cdb457ffce692a0fb54a4881281
  Stored in directory: c:\users\naksh\appdata\local\pip\cache\wheels\20\91\57\2965d4cff1b8ac7ed1b6fa25741882af3974b54a31759e10b6
Successfully built scikit-surprise
Note: you may need to restart the kernel to use updated packages.


In [45]:
super_df['product'].nunique()

27

In [44]:
super_df['author'].nunique()

7

### Collaborative Filtering Methods

In [7]:
from surprise import Dataset,Reader

reader = Reader(rating_scale=(1, 10))
data = Dataset.load_from_df(super_df[['author', 'product', 'score']], reader)
data

In [8]:
from surprise.model_selection import train_test_split
trainset, testset = train_test_split(data, test_size=.25,random_state=123)

In [48]:
from surprise import SVD, KNNWithMeans, accuracy

In [49]:
svd_model = SVD(n_factors=5,biased=False)
svd_model.fit(trainset)

In [52]:
test_pred = svd_model.test(testset)

In [53]:
# compute RMSE
accuracy.rmse(test_pred)

RMSE: 2.8928


2.892763522325869

In [64]:
from surprise.model_selection import cross_validate
cross_itembased = cross_validate(svd_model, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    2.9409  2.9358  2.8974  2.9245  2.8652  2.9127  0.0282  
MAE (testset)     2.5376  2.5353  2.4911  2.5422  2.3765  2.4965  0.0628  
Fit time          0.65    0.37    0.38    0.28    0.31    0.40    0.13    
Test time         0.03    0.04    0.02    0.03    0.03    0.03    0.01    


In [54]:
# Item based KNNWithMeans

In [55]:
knn_i = KNNWithMeans(k=10, sim_options={ 'user_based': False})
knn_i.fit(trainset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [57]:
test_pred=knn_i.test(testset)
accuracy.rmse(test_pred)

RMSE: 2.9349


2.934891045689189

In [62]:
from surprise.model_selection import cross_validate
cross_itembased = cross_validate(knn_i, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE, MAE of algorithm KNNWithMeans on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    2.9606  3.0109  2.9797  2.9990  3.0113  2.9923  0.0196  
MAE (testset)     2.4575  2.4551  2.4300  2.4224  2.4390  2.4408  0.0137  
Fit time          6.19    6.06    6.07    5.88    21.83   9.21    6.31    
Test time         13.04   12.42   12.53   39.64   35.27   22.58   12.23   


In [58]:
# User based KNNWithMeans

In [59]:
knn_u = KNNWithMeans(k=10, sim_options={ 'user_based': True})
knn_u.fit(trainset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [60]:
test_pred=knn_u.test(testset)
accuracy.rmse(test_pred)

RMSE: 2.9351


2.9350713028359814

In [63]:
from surprise.model_selection import cross_validate
cross_itembased = cross_validate(knn_u, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE, MAE of algorithm KNNWithMeans on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    2.9683  2.9234  3.0338  2.9408  2.9175  2.9567  0.0424  
MAE (testset)     2.4143  2.2443  2.3499  2.2542  2.2502  2.3026  0.0681  
Fit time          1.16    3.02    2.55    4.35    1.60    2.53    1.12    
Test time         3.56    5.03    4.89    3.66    3.92    4.21    0.62    


### Here we see that the SVD is the best collaborative filtering model for the recommendation
### as compared to KNNWithMeans item and user based

In [69]:
test_pred = svd_model.test(testset)

In [71]:
test_pred

[Prediction(uid='Amazon Customer', iid='OnePlus 3 (Graphite, 64 GB)', r_ui=10.0, est=9.098155287881031, details={'was_impossible': False}),
 Prediction(uid='Amazon Customer', iid='Lenovo Vibe K5 (Silver, 16GB)', r_ui=2.0, est=7.087152889143062, details={'was_impossible': False}),
 Prediction(uid='Amazon Customer', iid='Motorola Moto G Turbo (White, 16GB)', r_ui=2.0, est=6.54367132981044, details={'was_impossible': False}),
 Prediction(uid='Cliente Amazon', iid="Huawei P9 Lite Smartphone, LTE, Display 5.2'' FHD, Processore Octa-Core Kirin 650, 16 GB Memoria Interna, 3GB RAM, Fotocamera 13 MP, Single-SIM, Android 6.0 Marshmallow, Bianco [Italia]", r_ui=10.0, est=9.307028883853636, details={'was_impossible': False}),
 Prediction(uid='Amazon Customer', iid='Lenovo Vibe K4 Note (White,16GB)', r_ui=6.0, est=7.0317505735591235, details={'was_impossible': False}),
 Prediction(uid='Amazon Customer', iid='Lenovo Vibe K4 Note (Black, 16GB)', r_ui=2.0, est=6.863114563262318, details={'was_impossib

In [70]:
t_pred = pd.DataFrame(test_pred)
t_pred

uid                                                iid  \
0     Amazon Customer                        OnePlus 3 (Graphite, 64 GB)   
1     Amazon Customer                      Lenovo Vibe K5 (Silver, 16GB)   
2     Amazon Customer                Motorola Moto G Turbo (White, 16GB)   
3      Cliente Amazon  Huawei P9 Lite Smartphone, LTE, Display 5.2'' ...   
4     Amazon Customer                   Lenovo Vibe K4 Note (White,16GB)   
...               ...                                                ...   
3554  Amazon Customer                  Lenovo Vibe K4 Note (Black, 16GB)   
3555  Amazon Customer                        OnePlus 3 (Graphite, 64 GB)   
3556  Amazon Customer                  Lenovo Vibe K4 Note (Black, 16GB)   
3557  Amazon Customer                Motorola Moto G Turbo (White, 16GB)   
3558  Amazon Customer                   Lenovo Vibe K4 Note (White,16GB)   

      r_ui       est                    details  
0     10.0  9.098155  {'was_impossible': False}  
1      2.0  7.087153  {'was_impossible': False}  
2      2.0  6.543671  {'was_impossible': False}  
3     10.0  9.307029  {'was_impossible': False}  
4      6.0  7.031751  {'was_impossible': False}  
...    ...       ...                        ...  
3554   4.0  6.863115  {'was_impossible': False}  
3555  10.0  9.098155  {'was_impossible': False}  
3556   8.0  6.863115  {'was_impossible': False}  
3557   8.0  6.543671  {'was_impossible': False}  
3558   2.0  7.031751  {'was_impossible': False}  

[3559 rows x 5 columns]

### The average predicted score for each user in the testset

In [113]:
avg_predicted_userscore = t_pred.groupby('uid').agg({'r_ui':'mean','est':'mean'}).reset_index()
avg_predicted_userscore

uid      r_ui       est
0  Amazon Customer  7.053061  7.113751
1        Anonymous  5.714286  6.164026
2       Anonymous   9.343750  9.452242
3   Cliente Amazon  8.509554  8.898118
4     einem Kunden  9.428571  9.466755
5     einer Kundin  9.777778  9.684926

In [80]:
avg_predicted_userscore['uid']

0    Amazon Customer
1          Anonymous
2         Anonymous 
3     Cliente Amazon
4       einem Kunden
5       einer Kundin
Name: uid, dtype: object

### Note : there seems to be 2 kinds of Anonymous in the user list , as we can see from above. Both are treated differently

### Q7. The SVD model has predicted the average score of the users and the average scores of the users seem to be almost equal to the true rating of the users. Here since we have taken users who have given at least 1000 ratings and products which have atleast 1000 ratings we are getting a more realistic understanding of user behaviour and preference. The model was able to predict that the users taste and it almost matches the true value of the user rating.  

In [75]:
tset = pd.DataFrame(testset,columns=['author','product','score'])
tset

author                                            product  \
0     Amazon Customer                        OnePlus 3 (Graphite, 64 GB)   
1     Amazon Customer                      Lenovo Vibe K5 (Silver, 16GB)   
2     Amazon Customer                Motorola Moto G Turbo (White, 16GB)   
3      Cliente Amazon  Huawei P9 Lite Smartphone, LTE, Display 5.2'' ...   
4     Amazon Customer                   Lenovo Vibe K4 Note (White,16GB)   
...               ...                                                ...   
3554  Amazon Customer                  Lenovo Vibe K4 Note (Black, 16GB)   
3555  Amazon Customer                        OnePlus 3 (Graphite, 64 GB)   
3556  Amazon Customer                  Lenovo Vibe K4 Note (Black, 16GB)   
3557  Amazon Customer                Motorola Moto G Turbo (White, 16GB)   
3558  Amazon Customer                   Lenovo Vibe K4 Note (White,16GB)   

      score  
0      10.0  
1       2.0  
2       2.0  
3      10.0  
4       6.0  
...     ...  
3554    4.0  
3555   10.0  
3556    8.0  
3557    8.0  
3558    2.0  

[3559 rows x 3 columns]

### Recommending the top 5 products to the users

In [93]:
pop5 = popularity_recommendations.head()

In [94]:
top5pred_df = t_pred[(t_pred['iid'].isin(pop5['product']))]

In [100]:
top5pred_df

uid                                                iid  r_ui  \
14    einer Kundin  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  10.0   
17    einer Kundin  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  10.0   
23    einer Kundin  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  10.0   
31      Anonymous               Samsung Galaxy S7 edge 32GB (Verizon)  10.0   
41      Anonymous                    Samsung Galaxy S7 32GB (Verizon)  10.0   
...            ...                                                ...   ...   
3478    Anonymous              Samsung Galaxy S7 edge 32GB (T-Mobile)   6.0   
3499    Anonymous                    Samsung Galaxy S7 32GB (Verizon)   8.0   
3500  einer Kundin  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  10.0   
3507  einer Kundin  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  10.0   
3530  einem Kunden  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  10.0   

           est                    details  
14    9.684926  {'was_impossible': False}  
17    9.684926  {'was_impossible': False}  
23    9.684926  {'was_impossible': False}  
31    9.320163  {'was_impossible': False}  
41    9.642848  {'was_impossible': False}  
...        ...                        ...  
3478  9.289920  {'was_impossible': False}  
3499  9.642848  {'was_impossible': False}  
3500  9.684926  {'was_impossible': False}  
3507  9.684926  {'was_impossible': False}  
3530  9.466755  {'was_impossible': False}  

[195 rows x 5 columns]

In [85]:
popularity_recommendations.head(10) 

product     score  Rank
21                  Samsung Galaxy S7 32GB (T-Mobile)  9.647059   1.0
25  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  9.580029   2.0
23             Samsung Galaxy S7 edge 32GB (T-Mobile)  9.527273   3.0
24              Samsung Galaxy S7 edge 32GB (Verizon)  9.437500   4.0
22                   Samsung Galaxy S7 32GB (Verizon)  9.414634   5.0
20                   Samsung Galaxy S5 16GB (Verizon)  9.205882   6.0
2   Huawei P9 Lite Smartphone, LTE, Display 5.2'' ...  9.043210   7.0
3                                            LG GX200  9.000000   8.0
4   Lenovo Motorola Moto G Smartphone, 4,5 pollici...  8.636364   9.0
16                        OnePlus 3 (Graphite, 64 GB)  8.530739  10.0

In [95]:
users_test = top5pred_df['uid'].unique()
users_test

array(['einer Kundin', 'Anonymous ', 'einem Kunden'], dtype=object)

In [109]:
top5pred_df[top5pred_df['uid'] == 'einer Kundin' ][['iid','r_ui','est']].drop_duplicates()

iid  r_ui       est
14    Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  10.0  9.684926
332   Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...   8.0  9.684926
1413  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...   6.0  9.684926
3000  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...   4.0  9.684926

In [111]:
for i in users_test:
    print('User name : ',i)
    print(top5pred_df[top5pred_df['uid'] == i ][['iid','r_ui','est']].drop_duplicates())
    print('---------------------------------------------------------------------')

User name :  einer Kundin
                                                    iid  r_ui       est
14    Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...  10.0  9.684926
332   Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...   8.0  9.684926
1413  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...   6.0  9.684926
3000  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...   4.0  9.684926
---------------------------------------------------------------------
User name :  Anonymous 
                                         iid  r_ui       est
31     Samsung Galaxy S7 edge 32GB (Verizon)  10.0  9.320163
41          Samsung Galaxy S7 32GB (Verizon)  10.0  9.642848
358         Samsung Galaxy S7 32GB (Verizon)   8.0  9.642848
417        Samsung Galaxy S7 32GB (T-Mobile)   9.0  9.755371
593    Samsung Galaxy S7 edge 32GB (Verizon)   9.0  9.320163
610   Samsung Galaxy S7 edge 32GB (T-Mobile)  10.0  9.289920
677        Samsung Galaxy S7 32GB (T-Mobile)  10.0  9.755371
813   Samsung Galaxy S7 edge 32G

### Above we have recommended the top 5 products to the test users and we observe their predictions as compared to their true ratings.

### Q10. In what business scenario you should use popularity based Recommendation Systems?

Popularity based recommender system is used when the user requires the most popular items, news, and information.
In a case where the user is looking for a Television, and the user would like suggestions - then the most popularly bought 
TV would be recommended based on the purchase history, the reviews and how many users have clicked on it. 

This form of recommendation is not personalized, and doesn't suffer from cold start.

### Q11. In what business scenario you should use CF based Recommendation Systems?

Collaborative Filtering based Recommender systems are used when the user requires personalized suggestions for the service or product they require. Let us consider that the user wants to watch a movie; here if the system understands the user's taste in movies, then it can recommend movies similar to his preference. CF based recommender systems do the very the same, by understanding what the user's preference is and suggesting products or services which they would like.

This system suffers from cold start issue whenever a new user or a new item is considered.

### Q12. What other possible methods can you think of which can further improve the recommendation for different users ? 

The system could be made a lot more enhanced if we understood:

    a. the Budget range of the users ( as different users have different budgets when buying a phone )
    b. the Brand preference of the users ( users may show affinity towards certain brands )